src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,426 @@
+            // starting with a node containing delimiters assumes the caret is placed inside those delimiters. 
+            // This covers cases like `obj.ToString$()`, where `token` references `(` but the caret isn't actually 
+            // inside the argument list.
+            if (token.IsKind(SyntaxKind.OpenBraceToken, SyntaxKind.OpenBracketToken, SyntaxKind.OpenParenToken)
+                && token.Span.Start >= caretPosition)
+            {
+                currentNode = currentNode.Parent;
+            }
+
+            if (currentNode == null)
+            {
+                return;
+            }
+
+            var isCaretAtEndOfLine = ((SnapshotPoint)caret).Position == ((SnapshotPoint)caret).GetContainingLine().End;
+            var syntaxFacts = document.GetLanguageService<ISyntaxFactsService>();
+            if (!LooksLikeNodeInArgumentList(currentNode, caret, isCaretAtEndOfLine, syntaxFacts))
+            {
+                return;
+            }
+
+            // verify all delimeters exist until you reach statement syntax that requires a semicolon
+            while (!IsStatementOrFieldDeclaration(currentNode, syntaxFacts))
+            {
+                if (!ClosingDelimiterExistsIfNeeded(currentNode))
+                {
+                    // A required delimiter is missing; do not treat semicolon as statement completion
+                    return;
+                }
+
+                if (currentNode.Parent == null)
+                {
+                    return;
+                }
+
+                currentNode = currentNode.Parent;
+            }
+
+            // if the statement syntax itself requires a closing delimeter, verify it is there
+            if (StatementClosingDelimiterIsMissing(currentNode))
+            {
+                // Example: missing final `)` in `do { } while (x$$`
+                return;
+            }
+
+            var semicolonPosition = GetSemicolonLocation(root, currentNode, caretPosition);
+
+            // Place cursor after the statement
+            args.TextView.TryMoveCaretToAndEnsureVisible(args.SubjectBuffer.CurrentSnapshot.GetPoint(semicolonPosition));
+        }
+
+        private int GetSemicolonLocation(SyntaxNode root, SyntaxNode currentNode, int caretPosition)
+        {
+            int end;
+            if (currentNode.IsKind(SyntaxKind.ForStatement))
+            {
+                // in for statements, semicolon can go after initializer or after condition, depending on where the caret is located
+                var forStatementSyntax = (ForStatementSyntax)currentNode;
+                if (CaretIsInForStatementCondition(caretPosition, forStatementSyntax))
+                {
+                    end = forStatementSyntax.Condition.FullSpan.End;
+                }
+                else if (CaretIsInForStatementDeclaration(caretPosition, forStatementSyntax))
+                {
+                    end = forStatementSyntax.Declaration.FullSpan.End;
```

Comment:
you are checking the 'Declaration' case, but you also need to check the 'Initializers' case.  This happens when you rite:

```for (i = 0, j = 1; ...```

i.e. there is no `int i = 0` just `i = 0`.

Reply:
I added tests for initializers and as a result found this issue https://github.com/dotnet/roslyn/issues/32250